src/rust/engine/fs/src/snapshot.rs

```diff
@@ -119,22 +116,69 @@ least one path component, but at least one did not"
         directory.set_directories(protobuf::RepeatedField::from_vec(dirs));
         directory.set_files(protobuf::RepeatedField::from_vec(files));
         let digest = store.record_directory(&directory)?;
-        Ok(Snapshot {
-          fingerprint: digest.0,
-          digest: Some(digest),
-          path_stats: path_stats,
-        })
+        Ok(Snapshot { digest, path_stats })
       })
       .to_boxed()
   }
+
+  // TODO: How much should we be aiming to parallelise this?
+  // The old method returned a Future which did all of the work in one work-unit, on the PosixFS
+  // CpuPool. Is that sufficient? Or should we be kicking off each LMDB read in parallel on the
+  // pool? I can start investigating the performance of parallel LMDB reads on APFS, but I wanted to
+  // check whether it was even worth thinking about before doing so.
+  //
+  // (This is currently not done in a pool at all because I want to resolve the TODO on PosixFS's
+  // pool() method before doing so).
+  pub fn contents(self, store: Arc<Store>) -> BoxFuture<Vec<FileContent>, String> {
+    futures::future::done(Snapshot::contents_for_directory_sync(
+      &self.digest.0,
+      store,
+      PathBuf::from(""),
+    )).to_boxed()
+  }
+
+  fn contents_for_directory_sync(
+    fingerprint: &Fingerprint,
+    store: Arc<Store>,
+    path_so_far: PathBuf,
+  ) -> Result<Vec<FileContent>, String> {
+    let directory = store.load_directory_proto(fingerprint)?.ok_or_else(|| {
+      format!("Could not find snapshot {}", fingerprint)
+    })?;
+
+    let mut contents = Vec::new();
+
+    for file_node in directory.get_files() {
+      contents.push(FileContent {
+        path: path_so_far.join(file_node.get_name()),
+        content: store
+          .load_file_bytes(&Fingerprint::from_hex_string(
+            file_node.get_digest().get_hash(),
+          ).unwrap())?
```

Comment:
The unwraps in here should be handled, as they would indicate IO errors or someone having deleted the store out from under a running process (both of which become more likely with the daemon).

Reply:
...or maybe not...? I suppose a panic will be somewhat-properly handled, and would trigger the daemon killing itself with a log message. But the client connected to the daemon would not get a useful message. If you were to trigger an error here saying something like "internal error, please run `./pants kill-pantsd`", the response is: "gee thanks, why didn't you just do that yourself?".

Six of one, half dozen of the other.